internal/decision/scoreledger.go

```diff
@@ -0,0 +1,349 @@
+
+			// Update the long-term score.
+			if updateLong {
+				if l.lastExchange.After(lastLongUpdate) {
+					l.longScore = ewma(l.longScore, longTermScore, longTermAlpha)
+				} else {
+					l.longScore = ewma(l.longScore, 0, longTermAlpha)
+				}
+			}
+
+			// Calculate the new score.
+			//
+			// The accounting score adjustment prefers peers _we_
+			// need over peers that need us. This doesn't help with
+			// leeching.
+			var lscore float64
+			if l.bytesRecv == 0 {
+				lscore = 0
+			} else {
+				lscore = float64(l.bytesRecv) / float64(l.bytesRecv+l.bytesSent)
+			}
+			score := int((l.shortScore + l.longScore) * (lscore*.5 + .75))
+
+			// Avoid updating the connection manager unless there's a change. This can be expensive.
+			if l.score != score {
+				// put these in a list so we can perform the updates outside _global_ the lock.
+				updates = append(updates, update{l.partner, score})
+				l.score = score
+			}
+			l.lock.Unlock()
+		}
+		dsl.lock.Unlock()
+
+		// record the times.
+		lastShortUpdate = now
+		if updateLong {
+			lastLongUpdate = now
+		}
+
+		// apply the updates
+		for _, update := range updates {
+			dsl.scorePeer(update.peer, update.score)
+		}
+		// Keep the memory. It's not much and it saves us from having to allocate.
+		updates = updates[:0]
+
+		// Used by the tests
+		if dsl.sampleCh != nil {
+			dsl.sampleCh <- struct{}{}
+		}
+	}
+}
+
+// Returns the score ledger for the given peer or nil if that peer
+// is not on the ledger.
+func (dsl *DefaultScoreLedger) find(p peer.ID) *scoreledger {
+	// Take a read lock (as it's less expensive) to check if we have
+	// a ledger for the peer.
+	dsl.lock.RLock()
+	l, ok := dsl.ledgerMap[p]
+	dsl.lock.RUnlock()
+	if ok {
+		return l
+	}
+	return nil
+}
+
+// Returns a new scoreledger.
+func newScoreLedger(p peer.ID) *scoreledger {
+	return &scoreledger{
+		partner: p,
+	}
+}
+
+// Lazily instantiates a ledger.
+func (dsl *DefaultScoreLedger) findOrCreate(p peer.ID) *scoreledger {
+	l := dsl.find(p)
+	if l != nil {
+		return l
+	}
+
+	// There's no ledger, so take a write lock, then check again and
+	// create the ledger if necessary.
+	dsl.lock.Lock()
+	defer dsl.lock.Unlock()
+	l, ok := dsl.ledgerMap[p]
+	if !ok {
+		l = newScoreLedger(p)
+		dsl.ledgerMap[p] = l
+	}
+	return l
+}
+
+// GetReceipt returns aggregated data communication with a given peer.
+func (dsl *DefaultScoreLedger) GetReceipt(p peer.ID) *Receipt {
+	l := dsl.find(p)
+	if l != nil {
+		return l.Receipt()
+	} else {
+		// Return a blank receipt.
+		return &Receipt{
+			Peer:      p.String(),
+			Value:     0,
+			Sent:      0,
+			Recv:      0,
+			Exchanged: 0,
+		}
+	}
+}
+
+// Starts the default ledger sampling process.
+func (dsl *DefaultScoreLedger) Start() {
+	go dsl.scoreWorker()
+}
+
+// Closes (stops) the sampling process.
+func (dsl *DefaultScoreLedger) Close() {
+	close(dsl.closing)
+}
+
+// Initializes the score ledger.
+func (dsl *DefaultScoreLedger) Init(scorePeer ScorePeerFunc) {
```

Comment:
Rather than having a separate `Init()` method, can we just create the ledger map and set the `ScorePeerFunc` in `NewDefaultScoreLedger()`? (It seems like ledgerMap is already initialized there)

Reply:
Hm... But we need it for abstraction. `Init()` is called here:

```go
func (e *Engine) initScoreLedger() {
	if e.scoreLedger == nil {
		e.scoreLedger = NewDefaultScoreLedger()
	}
	e.scoreLedger.Init(func(p peer.ID, score int) {
		if score == 0 {
			e.peerTagger.UntagPeer(p, e.tagUseful)
		} else {
			e.peerTagger.TagPeer(p, e.tagUseful, score)
		}
	})
}
```
`NewDefaultScoreLedger()` is only used by default. But we need to pass _our_ function to any `ScoreLedger`.

Reply:
I see, in that case I would suggest
- remove the `Init()` method
- add `ScorePeerFunc` to the `Start()` method instead: `Start(ScorePeerFunc)`